File: Chapter04/ReflectedColors/ReflectedColors/ReflectedColors/ReflectedColorsPage.cs
using System;
using System.Reflection;
using Xamarin.Forms;

namespace ReflectedColors
{
    public class ReflectedColorsPage : ContentPage
    {
        public ReflectedColorsPage()
        {
            StackLayout stackLayout = new StackLayout();

            // Loop through the Color structure fields.
            foreach (FieldInfo info in typeof(Color).GetRuntimeFields())
            {
                // Skip the obsolete (i.e. misspelled) cplors.
                if (info.GetCustomAttribute<ObsoleteAttribute>() != null) 
                    continue;

                if (info.IsPublic &&
                    info.IsStatic &&
                    info.FieldType == typeof(Color))
                {
                    stackLayout.Children.Add(
                        CreateColorLabel((Color)info.GetValue(null), info.Name));
                }
            }

            // Loop through the Color structure properties.
            foreach (PropertyInfo info in typeof(Color).GetRuntimeProperties())
            {
                MethodInfo methodInfo = info.GetMethod;

                if (methodInfo.IsPublic &&
                    methodInfo.IsStatic &&
                    methodInfo.ReturnType == typeof(Color))
                {
                    stackLayout.Children.Add(
                        CreateColorLabel((Color)info.GetValue(null), info.Name));
                }
            }

            Padding = new Thickness(5, Device.RuntimePlatform == Device.iOS ? 20 : 5, 5, 5);

            // Put the StackLayout in a ScrollView.
            Content = new ScrollView
            {
                Content = stackLayout
            };
        }

        Label CreateColorLabel(Color color, string name)
        {
            Color backgroundColor = Color.Default;

            if (color != Color.Default)
            {
                // Standard luminance calculation
                double luminance = 0.30 * color.R +
                                   0.59 * color.G +
                                   0.11 * color.B;

                backgroundColor = luminance > 0.5 ? Color.Black : Color.White;
            }

            // Create the Label.
            return new Label
            {
                Text = name,
                TextColor = color,
                FontSize = Device.GetNamedSize(NamedSize.Large, typeof(Label)),
                BackgroundColor = backgroundColor
            };
        }
    }
}

File: Chapter04/FS/ReflectedColors/ReflectedColors/ReflectedColors/ReflectedColorsPage.fs
namespace ReflectedColors

open System
open System.Reflection
open Xamarin.Forms

type ReflectedColorsPage() = 
    inherit ContentPage()

    let stackLayout = StackLayout()

    // Define function to add Label to StackLayout to display color.
    let AddToStack = fun color name ->
        let backgroundColor = 
            if color = Color.Default then Color.Default
            else 
                // Standard luminance calculation.
                let luminance = 0.30 * color.R +
                                0.59 * color.G +
                                0.11 * color.B 
                if luminance > 0.5 then Color.Black else Color.White

        // Create the Label and add to StackLayout.
        stackLayout.Children.Add(Label(Text = name,
                                       TextColor = color,
                                       FontSize = Device.GetNamedSize(NamedSize.Large, typeof<Label>),
                                       BackgroundColor = backgroundColor)) 

    // Loop through the Color structure fields.
    do for info in typeof<Color>.GetRuntimeFields() do
        if info.IsPublic && 
           info.IsStatic && 
           info.FieldType = typeof<Color> && 
           info.GetCustomAttribute<ObsoleteAttribute>() = null  // Skip the obsolete (i.e. misspelled) colors
        then AddToStack (info.GetValue(null) :?> Color) info.Name
        
    // Loop through the Color structure properties.
    do for info in typeof<Color>.GetRuntimeProperties() do
        let methodInfo = info.GetMethod
        if methodInfo.IsPublic &&
           methodInfo.IsStatic &&
           methodInfo.ReturnType = typeof<Color>
        then AddToStack (info.GetValue(null) :?> Color) info.Name

    do base.Padding <- match Device.RuntimePlatform with 
                            | Device.iOS -> Thickness(5.0, 20.0, 5.0, 5.0)
                            | _ -> Thickness(5.0, 5.0, 5.0, 5.0)

    // Put the StackLayout in a ScrollView.
    do base.Content <- ScrollView(Content = stackLayout)
